### PR TITLE
268725 anz enhancement address validation demo engine

### PR DESCRIPTION
**Pull request by sufangl**

### CODE REVIEW

File: src/ts/address-search.ts [hunk 1]
```diff
@@ -1334,7 +1436,8 @@ export default class AddressValidation {
     },
     // Decide whether to either show a picklist or a verified result from a Validate response
     handleValidateResponse: (response: SearchResponse) => {
-      if (response.result.confidence === 'Verified match') {
+      // todo: sufang to add confidence === interaction required?? to confirm
+      if (response.result.confidence === 'Verified match' || response.result.confidence === 'Interaction required') {
```
[the remark this answers is not in the document]
Comment (sufangl) [reply src/ts/address-search.ts]: Introduced AddressValidationConfidenceType enum

File: src/ts/address-search.ts [hunk 2]
```diff
@@ -251,9 +317,27 @@ export default class AddressValidation {
...
+      }
+      this.countryDropdown.sort((a, b) => a.country.localeCompare(b.country))
+      this.events.trigger('post-datasets-update');
+    }
+  }
+
   // When a country from the list is changed, update the current country code, call the promptset endpoint again and reset to the default search mode
   private handleCountryListChange(): void {
     this.currentCountryCode = this.options.elements.countryList.value;
+    this.currentCountryName = this.options.elements.countryList[this.options.elements.countryList.selectedIndex].label;
```
[the remark this answers is not in the document]
Comment (sufangl) [reply src/ts/address-search.ts]: Updated

File: src/ts/address-search.ts [hunk 3]
```diff
@@ -55,19 +82,52 @@ export default class AddressValidation {
 
   public getEnrichmentData(globalAddressKey: string) {
     if (globalAddressKey) {
-      const data = {
+      var regionalAttributes: {};
+      var premium_location_insight: {}
```
[the remark this answers is not in the document]
Comment (sufangl) [reply src/ts/address-search.ts]: Updated

File: src/ts/address-search.ts [hunk 4]
```diff
@@ -55,19 +82,52 @@ export default class AddressValidation {
 
   public getEnrichmentData(globalAddressKey: string) {
     if (globalAddressKey) {
-      const data = {
+      var regionalAttributes: {};
```
[the remark this answers is not in the document]
Comment (sufangl) [reply src/ts/address-search.ts]: Updated

File: src/ts/address-search.ts [hunk 5]
```diff
@@ -12,7 +30,15 @@ export default class AddressValidation {
   public events;
   public request: Request;
 
+  public countryDropdown: any[] = [];
```
[the remark this answers is not in the document]
Comment (sufangl) [reply src/ts/address-search.ts]: Updated

File: index.html [hunk 6]
```diff
@@ -333,6 +96,10 @@ <h5 class="header">Enter partial address, then get results</h5>
           <label for="country">Country</label>
           <input type="text" name="country" id="country">
         </div>
+
+        <div class="error-display hidden">
+          <label style="color:red; font-weight: bold;"></label>
```
[the remark this answers is not in the document]
Comment (sufangl) [reply index.html]: Updated

File: index.html [hunk 7]
```diff
@@ -389,6 +165,26 @@ <h2>Validated address information</h2>
         }
       }
 
+      // populate the country dataset dropdown with the authorized country datasets
+      address.events.on("post-datasets-update", function() {
+        let countryListElement = options.elements.countryList;
+        let optionElements = countryListElement.getElementsByTagName("option"), index
```
[the remark this answers is not in the document]
Comment (sufangl) [reply index.html]: Removed unnecessary index

File: index.html [hunk 8]
```diff
@@ -571,8 +401,52 @@ <h2>Validated address information</h2>
...
+
+          collectionMap.forEach((value, key) => {
+            const htmlSpanElement = document.createElement("span");
+            let htmlBrElement = document.createElement("br");
+            if (typeof value == "object") {
+              htmlSpanElement.innerText = key + ": ";
+              collectionParentElement.append(htmlSpanElement, htmlBrElement);
+              addChildElement(Object.entries(value), collectionParentElement, 1)
+            } else {
+              htmlSpanElement.innerText = key + ": " + value;
```
[the remark this answers is not in the document]
Comment (sufangl) [reply index.html]: Updated

File: index.html [hunk 9]
```diff
@@ -571,8 +401,52 @@ <h2>Validated address information</h2>
...
+          });
+          collectionParentElement.append(document.createElement("br"));
+        }
+      }
+
+      function addChildElement(entries, componentCollectionsElement, level) {
+        for (const [childKey, childValue] of entries) {
+          const htmlChildSpanElement = document.createElement("span");
+          let htmlChildBrElement = document.createElement("br");
+          htmlChildSpanElement.setAttribute("class", "tab");
```
[the remark this answers is not in the document]
Comment (sufangl) [reply index.html]: Updated

**Review by jbate — APPROVED** [review]
Well done @sufangl - lots of good stuff here. I've added some minor comments that you may want to address before merging.

File: src/ts/address-search.ts [hunk 10]
```diff
@@ -414,25 +536,37 @@ export default class AddressValidation {
...
+          data = this.generateLookupDataForApiCall(this.currentSearchTerm, AddressValidationLookupKeywords.UDPRN.key);
+          url = this.baseUrl + this.lookupV2Endpoint;
           headers = [{ key: 'Add-Addresses', value: true }];
           callback = this.picklist.showLookup;
-          break; 
-        } 
+          break;
+        }
+        case AddressValidationMode.LOOKUPV2: {
+          const lookupSeacrhTerm = this.currentSearchTerm.split(',');
```
[the remark this answers is not in the document]
Comment (sufangl) [reply src/ts/address-search.ts:555]: Updated

File: src/ts/address-search.ts [hunk 11]
```diff
@@ -160,6 +243,16 @@ export default class AddressValidation {
           ];
           setTimeout(() => this.handlePromptsetResult({ result: { lines } }));
           return;
+        } else if (this.searchType === AddressValidationSearchType.LOOKUPV2) {
+          const lines = [
+            {prompt: 'Lookup type', suggested_input_length: 160,
+              dropdown_options: [AddressValidationLookupKeywords.LOCALITY, AddressValidationLookupKeywords.POSTAL_CODE]},
+            {prompt: 'Add addresses (If True concrete addresses is returned in response)',
```
[the remark this answers is not in the document]
Comment (sufangl) [reply src/ts/address-search.ts:250]: Updated

File: src/ts/address-search.ts [hunk 12]
```diff
@@ -25,7 +58,10 @@ export default class AddressValidation {
   private inputs: HTMLInputElement[];
   private lastSearchTerm: string;
   private currentSearchTerm: string;
+  private lookupType: string;
+  private returnAddresses: boolean = true;
```
[the remark this answers is not in the document]
Comment (sufangl) [reply src/ts/address-search.ts:62]: Updated

File: index.html [hunk 13]
```diff
@@ -571,8 +471,103 @@ <h2>Validated address information</h2>
...
         document.querySelector(".metadata #delivery-address-key").innerHTML = data.result.address ? '<img src="./dist/images/marker-icon-s.png"/>' : '';
         document.querySelector(".metadata #delivery-address-value").innerHTML = data.result.address ? Object.values(data.result.address).filter(line => line !== "").join("<br>") : '';
-
         document.querySelector(".metadata").classList.remove("invisible");
+
+        populateAddressAdditionalInfo(address.componentsCollectionMap, document.querySelector("#components-collection"));
+        populateAddressAdditionalInfo(address.metadataCollectionMap, document.querySelector("#metadata-collection"));
+      }
+
+      function populateAddressAdditionalInfo(collectionMap, parentElement, elementTitle, collapsibleIndex) {
```
[the remark this answers is not in the document]
Comment (sufangl) [reply index.html:480]: Added comment

File: index.html [hunk 14]
```diff
@@ -462,17 +328,38 @@ <h2>Validated address information</h2>
...
-        document.querySelector("button#find-address-button").classList[(address.searchType === "singleline" || address.searchType === "validate") ? 'remove' : 'add']("hidden");
+        document.querySelector("button#find-address-button").classList[
+                (address.searchType !== "autocomplete") ? 'remove' : 'add']("hidden");
+      });
+
+      // To display error when unsupported search type is selected
+      address.events.on("error-display", function (error) {
+        document.querySelector('.promptset').classList.add('hidden');
+        resetMetadata();
+        document.querySelector('.formatted-address').classList.add('hidden');
```
[the remark this answers is not in the document]
Comment (sufangl) [reply index.html:339]: Sorry for overlooked. Removed this line

File: index.html [hunk 15]
```diff
@@ -410,11 +237,33 @@ <h2>Validated address information</h2>
         populateMetadata(data);
       });
 
+      address.events.on("post-formatting-lookup", function(key, item) {
+        document.querySelector(".loader").classList.add("hidden");
+        document.querySelector("#validated-address-info").classList.add("hidden");
+
+        document.querySelector(".formatted-address").classList.remove("hidden");
```
[the remark this answers is not in the document]
Comment (sufangl) [reply index.html:244]: Sorry for overlooked. Removed this line

File: index.html [hunk 16]
```diff
@@ -389,6 +195,26 @@ <h2>Validated address information</h2>
         }
       }
 
+      // populate the country dataset dropdown with the authorized country datasets
+      address.events.on("post-datasets-update", function() {
+        let countryListElement = options.elements.countryList;
+        let optionElements = countryListElement.getElementsByTagName("option")
+        for (index = optionElements.length - 1; index >= 0; index--) {
```
[the remark this answers is not in the document]
Comment (sufangl) [reply index.html:202]: Updated

File: index.html [hunk 17]
```diff
@@ -333,15 +103,51 @@ <h5 class="header">Enter partial address, then get results</h5>
...
-        <div class="data-grid">
-          <span id="confidence-key"></span><span id="confidence-value"></span>
-          <span id="delivery-type-key"></span><span id="delivery-type-value"></span>
-          <span id="delivery-address-key"></span><span id="delivery-address-value"></span>
-          <span id="what3words-key" class="hidden"><img src="./dist/images/w3w.png"/></span><span id="what3words-value" class="hidden"></span>
+        <div class="validated-address-info" id="validated-address-info">
+          <h2>Validated address information</h2>
+          <span class="collapsible">
+            <span class="hide">[Hide]</span>
+            <span class="show hidden">[Show]</span>
```
[the remark this answers is not in the document]
Comment (sufangl) [reply index.html:117]: At this moment, these options are not keyboard accessible. 
Thanks for your suggestion but will put this as low priority item for now.

File: index.html
[the remark this answers is not in the document]
Comment (sufangl) [reply index.html:1]: Extracted the JavaScript to separate js files.

**Pull request merged**